File: LeetcodePractice/src/main/java/com/Bernie/leetcode/editor/cn/[59]螺旋矩阵 II.java
//给你一个正整数 n ，生成一个包含 1 到 n2 所有元素，且元素按顺时针顺序螺旋排列的 n x n 正方形矩阵 matrix 。
//
// 
//
// 示例 1： 
//
// 
//输入：n = 3
//输出：[[1,2,3],[8,9,4],[7,6,5]]
// 
//
// 示例 2： 
//
// 
//输入：n = 1
//输出：[[1]]
// 
//
// 
//
// 提示： 
//
// 
// 1 <= n <= 20 
// 
// Related Topics 数组 
// 👍 355 👎 0


package com.Bernie.leetcode.editor.cn;

class SpiralMatrixIi {
    public static void main(String[] args) {
        Solution solution = new SpiralMatrixIi().new Solution();
        int[][] res = solution.generateMatrix(3);
        for (int i = 0; i < res.length; i++) {
            for (int j = 0; j < res[0].length; j++) {
                System.out.print(res[i][j] + "  ");
            }
            System.out.println();
        }
    }

    //leetcode submit region begin(Prohibit modification and deletion)
    class Solution {
        public int[][] generateMatrix(int n) {
            if (n < 0) return new int[0][];
            int[][] ans = new int[n][n];
            int allNums = n * n;
            int left = 0, top = 0, right = n - 1, bottom = n - 1, num = 1;
            while (num <= allNums) {
                // left -> right
                for (int i = left; i <= right; i++) {
                    ans[left][i] = num++;
                }
                top++;
                // top->bottom
                for (int i = top; i <= bottom; i++) {
                    ans[i][right] = num++;
                }
                right--;
                // right -> left
                for (int i = right; i >= left; i--) {
                    ans[bottom][i] = num++;
                }
                bottom--;
                // bottom -> top
                for (int i = bottom; i >= top; i--) {
                    ans[i][left] = num++;
                }
                left++;

            }

            return ans;
        }
    }
//leetcode submit region end(Prohibit modification and deletion)

}

File: LeetcodePractice/src/main/java/com/Bernie/leetcode/editor/cn/[59]螺旋矩阵 II.py
from typing import List


class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        res = list()
        if n < 1 or n > 20:
            return res
        allNums = n * n
        res = [[0 for col in range(n)] for row in range(n)]
        left, top, right, bottom, num = 0, 0, n - 1, n - 1, 1
        while num <= allNums:
            # left -> right
            for column in range(left, right + 1):
                res[left][column] = num
                num = num + 1
            top = top + 1
            # top -> bottom
            for column in range(top, bottom + 1):
                res[column][right] = num
                num = num + 1
            right = right - 1
            # right -> left
            for column in range(right, left - 1, -1):
                res[bottom][column] = num
                num = num + 1
            bottom = bottom - 1
            # bottom ->  top
            for column in range(bottom, top - 1, -1):
                res[column][left] = num
                num = num + 1
            left = left + 1
        return res

solution = Solution()

multilist = solution.generateMatrix(3)

print(multilist)